docs/advanced-guides/testing/testing-with-react-testing-library.md

```diff
@@ -1,13 +1,117 @@
+
+A recommended simple test looks like the following:
+
+```jsx
+test('renders react router header', () => {
+  const { getByText } = render(<Router><App /></Router>);
+  const header = getByText('Welcome to React Router!');
+  expect(header).toBeInTheDocument();
+});
+```
+
+This ensures that we can render our `App` component  and that the `h1` we put in during our setup guide is in the document.
 
 ## Testing Routes and Redirects
 
 TODO
 
 ## Testing Links and Navigation
 
-TODO
+Testing a link and the subsequent navigation with React Testing Library is as easy as firing a click event on the link itself and asserting on the window's location to test that it worked.
+
+This is accomplished like so:
+
+```jsx
+it('goes to about when link clicked', () => {
+  const { getByText } = render(<Router><App /></Router>);
+  
+  const aboutLink = getByText('About');
+  act(() => {
+    fireEvent.click(aboutLink);
+  });
+
+  expect(window.location.pathname).toBe('/about');
+});
+```
+
+Since our setup guide then replaces the about link with a link back to the home page, we can easily test that we end up back on the home page by triggering a second click, this time on the home link:
+
+```jsx
+it('goes to about when link clicked and then back to home when link clicked', () => {
+  const { getByText } = render(<Router><App /></Router>);
+
+  const aboutLink = getByText('About');
+  act(() => {
+    fireEvent.click(aboutLink);
+  });
+  expect(window.location.pathname).toBe('/about');
+
+  const homeLink = getByText('Home');
+  act(() => {
+    fireEvent.click(homeLink);
+  });
+  expect(window.location.pathname).toBe('/');
+});
+```
+
+This test will initially fail if we run it right after the previous test. This is because the history stack still thinks we are on the about page. To fix this, we can use Jest's `beforeEach` function to replace our history state using: `window.history.replaceState({}, '', '/');`. This makes `react-router` think that we are on the home page when we start each test. Now, all three tests should be passing when ran together.
+
+Below is the full file to help see all of the tests together:
+
+```jsx
+import React from 'react';
+import { render, fireEvent, cleanup, act } from '@testing-library/react';
+import App from './App';
+import { BrowserRouter as Router } from "react-router-dom";
+
+beforeEach(() => {
+  window.history.replaceState({}, '', '/');
+});
+
+afterEach(cleanup);
+
+test('renders react router header', () => {
+  const { getByText } = render(<Router><App /></Router>);
+  const header = getByText('Welcome to React Router!');
+  expect(header).toBeInTheDocument();
+});
+
+it('goes to about when link clicked', () => {
+  const { getByText } = render(<Router><App /></Router>);
+  
+  const aboutLink = getByText('About');
+  act(() => {
```

Comment:
You shouldn't need to use `act` around `fireEvent` because [it's already wrapped in `act`](https://github.com/testing-library/react-testing-library/blob/04e64bb73063ae718b7fc819cdfb43e484f4ff6b/src/pure.js#L105-L126).

You should rarely need to use `act` in regular tests. For more on this: https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning